MIC: Introduce stock tracking hooks and AS scheduling for Back in Stock Notifications

xristos3490 opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR contributes to a larger initiative aimed at implementing back-in-stock notifications. It sets up the fundamental components of this feature, including the stock change event handler and the Action Scheduler's background task for broadcasting all notifications.

## Specification

### StockSyncController

This component manages changes in stock status events:

- Monitors `woocommerce_product_set_stock_status` and `woocommerce_variation_set_stock_status` hooks
- Validates products against supported types and eligible stock statuses
- Queues products for notification processing on shutdown

### NotificationsProcessor

This is the primary class for handling asynchronous notification broadcasts:

- Handles async batch processing of notifications via Action Scheduler
- Manages processing state using WordPress options (`wc_stock_notifications_cycle_state_*`)
- Implements configurable batch sizes, delays, and throttling mechanisms
- Tracks cycle metrics _(sent, failed, skipped counts, duration)_
- System-wide logger

### Variable Product & Variation Handling
The system handles variable product hierarchy:
- **Variable Products:** When a parent variable product stock changes, the system identifies child variations that inherit stock management from the parent using `StockManagementHelper::get_managed_variations()`
- **Variation Products:** Direct variation stock changes are handled independently

## The flow

![Screenshot 2025-05-30 at 4 45 42 PM](https://github.com/user-attachments/assets/9ad4b9ad-387d-489f-91c2-cb86d51cbe12)

## Open questions
- The identification of "non-managing-stock" variations within `StockManagementHelper::get_managed_variations()`, seems quite bold. I believe this approach is preferable to using `$variable->get_available_variations()`. The inspiration comes from the variable data store [here](https://github.com/woocommerce/woocommerce/blob/2ac4ec89d270321e8d24061fd9234e1f3d39057d/plugins/woocommerce/includes/data-stores/class-wc-product-variable-data-store-cpt.php#L586). There is definitely potential for introducing caching of variation attributes in a more generic way.

### Screenshots or screen recordings:

Here's a screen recording that showcases the broadcast process when a product with active notifications comes back in stock:

https://github.com/user-attachments/assets/a1b2d023-7d3a-4fb8-a238-ba8ea52a0cf6

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Prerequisites

##### 1. Insert some notifications

Install the following PHP snippet to add a new WP-CLI command for creating records (e.g., in your functions.php):

```php
add_action(
	'init',
	function () {
		if ( ! class_exists( '\WP_CLI' ) ) {
			return;
		}

		WP_CLI::add_command(
			'wc stock-notifications seed',
			function ( $args ) {
				$number     = $args[0];
				$product_id = $args[1];

				if ( $product_id < 1 ) {
					WP_CLI::error( 'Invalid product ID.' );
				}

				$product = wc_get_product( $product_id );

				if ( ! $product ) {
					WP_CLI::error( 'Invalid product.' );
				}

				$progress = \WP_CLI\Utils\make_progress_bar( 'Generating notifications', $number );

				$day = 0;
				for ( $i = 0; $i < $number; $i++ ) {

					$user_email = 'test' . wp_rand() . '@example.com';
					$args = array(
						'user_id'            => 0,
						'user_email'         => $user_email,
						'product_id'         => $product_id,
						'status'             => 'active',
					);

					$notification = new Automattic\WooCommerce\Internal\StockNotifications\Notification();
					$notification->set_props( $args );
					$re = $notification->save();
					if ( is_wp_error( $re ) ) {
						$progress->finish();
						WP_CLI::error( $re->get_error_message() );
					} else {
						$progress->tick();
					}
				}

				$progress->finish();

				WP_CLI::success( sprintf( 'Successfully seeded %d notifications for product ID %d.', $number, $product_id ) );
			}
		);
	}
);
```

Then, in your test site path, run:
```
wp wc stock-notifications seed <number> <product-id>
wp wc stock-notifications seed 1000 123
```

##### 2. Ensure the mail catcher is set up (e.g., Mailhog)

#### Test Scenario 1: Simple Product Stock Notification

##### Testing a Simple Product

1. Create or use an existing out-of-stock simple product.
2. Create notifications using the WP-CLI command above.
3. Edit the product and set the stock status to in-stock.
4. Go to WooCommerce > Status > Scheduled Actions.
5. Search for: `wc_send_stock_notifications_batch.`
6. Verify that a new action is scheduled (default: 60 seconds delay).
7. Note that the action arguments include the product ID.
8. Navigate to WooCommerce > Status > Logs.
9. Select log source: `wc-stock-notifications.`
10. Check the logs:
```
INFO: Scheduled stock notification for product 123
INFO: Completed cycle for product 123. Sent: 15, Skipped: 2, Failed: 0, Duration: 8 seconds. Total notifications processed: 17
```

##### Repeat with a Single Product that Manages Stock
1. Start with a stock quantity of 0.
2. Register notifications via WP-CLI.
3. Edit the product and set the stock quantity to 5.
4. Repeat steps 6-10 from above.

#### Test Scenario 2: Variable Product with Variations
Setup:
- Create an out-of-stock variable product with 2 variations:
  - Parent: "Test Variable Product"
  - Variation 1: Size Small (inherits stock from parent)
  - Variation 2: Size Large (stock: 0, out-of-stock)
Create Notifications:
- Create notifications via WP-CLI for the Small variation specifically.
Trigger Stock Change:
- Edit the variable product and set the stock status to in-stock.
- Save the changes.
Monitor Processing:
- Check Scheduled Actions for `wc_send_stock_notifications_batch.`
- Verify the action includes the parent product ID.
- Monitor logs for variable product processing messages.
- Confirm notifications for the variation are handled successfully.
- Now add notifications for the Large variation.
- Set the stock quantity of the Large variation to 5.
- Check again under Scheduled Actions for `wc_send_stock_notifications_batch.`
- Verify the action includes the variation product ID.
- Ensure notifications are handled as expected.

#### Test Scenario 3: Error Handling
Setup:
- Create an out-of-stock product with active notifications.
- Set the stock status to in-stock.
- Delete the product before the AS job starts.
- Once this runs, check the logs for error handling:
```
ERROR: Background process for product 123 terminated. Reason: Product 123 not found.
``` 

> [!NOTE]  
> Monitor for queries can be done by running the following commands:
> - `touch /Users/myuser/mysql_bin.log`
>
> Now within the MySQL command line:
> - `SET global log_output = 'FILE';`
> - `SET global general_log_file='/Users/myuser/mysql_bin.log';`
> - `SET global general_log = 1;`
> 
> To disable logging:
> - `SET global general_log = 0;`
> 
> To monitor queries for BIS, exit the MySQL console and run:
> - `tail -n 400 -f ~/mysql_bin.log | grep -E 'wc_stock_notif|manage_stock'`
>
> _Be aware this file can be huge._

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

xristos3490 committed:
Add base service and database schema

xristos3490 committed:
Make linter happy

xristos3490 committed:
Linter ignore

xristos3490 committed:
Merge branch 'mic/back-in-stock-notifications' into mic/bis-tables

xristos3490 committed:
Rename datastore class

xristos3490 committed:
Add basic CRUD for notifications, metadata, and a helper for logs

xristos3490 committed:
Make linter happy again

xristos3490 committed:
Shorten table names, suffix dates with gmt, introduce dedicated meta store

xristos3490 committed:
Fix docblock on reusable datastore

xristos3490 committed:
Lint fixes

xristos3490 committed:
Lint fixes

xristos3490 committed:
Update service provider

xristos3490 committed:
Lint fixes

xristos3490 committed:
Lint fixes

xristos3490 committed:
Lint fixes

xristos3490 committed:
Lint fixes

xristos3490 committed:
Fix dbDelta issues with capital letters and phpcs fixes

xristos3490 committed:
Add specific query formats

xristos3490 committed:
Remove object keyword refs

xristos3490 committed:
Remove meta cache duplication in favor of WC_Data level caching

xristos3490 committed:
Improvements in stock notifications CRUD and introduce WP_Error returns

xristos3490 committed:
Docblock improvements

xristos3490 committed:
Create the main controller and wire the data store

xristos3490 committed:
Fix data stores for basic CRUD and add unit tests

xristos3490 committed:
Add metadata handling and unit tests

xristos3490 committed:
Lint fixes

xristos3490 committed:
Organize unit tests and align create and modified dates

xristos3490 committed:
Alias main controller init and make linter happy

xristos3490 committed:
One more lint fix

xristos3490 committed:
Imporvements and new db table indexes

xristos3490 committed:
Clean up notification unit tests

xristos3490 committed:
Add metadata unit tests

xristos3490 committed:
Rename the event log helper

xristos3490 committed:
Add a factory, unit tests, and move notifications tests

xristos3490 committed:
Lint fixes

xristos3490 committed:
Make linter happy

xristos3490 committed:
Add a query method

xristos3490 committed:
Create data store for logs

xristos3490 committed:
Lint fixes

xristos3490 committed:
Lint fixes and fix the init data store of logs

xristos3490 committed:
Fix error in log params

xristos3490 committed:
Merge branch 'mic/back-in-stock-notifications' into mic/bis-tables

xristos3490 committed:
Use the main datastore for tests

xristos3490 committed:
Use specified date modified value and some polishing

xristos3490 committed:
Fix unit tests by applying changes on the object after saving the meta

xristos3490 committed:
Merge branch 'mic/back-in-stock-notifications' into mic/bis-tables

xristos3490 committed:
Lint fixes

xristos3490 committed:
Setup timezone for tests to control output

xristos3490 committed:
Avoid error logging in unit tests

xristos3490 committed:
Fix timezones when reading data from DB

xristos3490 committed:
Lint fixes

xristos3490 committed:
Reorg the table names into getters in data stores

xristos3490 committed:
Merge branch 'mic/back-in-stock-notifications' into mic/bis-tables

xristos3490 committed:
Minor fixes on the scheme

xristos3490 committed:
Move delete functions in each data store

xristos3490 committed:
Lint fixes

xristos3490 committed:
Fix unit tests

xristos3490 committed:
Remove comment ignores

xristos3490 committed:
Use assertCount

xristos3490 committed:
Fix unit tests

xristos3490 committed:
Add args and return types in activity log data store

xristos3490 committed:
Added args and return types in notifications data store where applicable

xristos3490 committed:
Fix the datastore docblock in activity logs tests

xristos3490 committed:
Differentiate dates for better coverage

xristos3490 committed:
Remove text based tests

xristos3490 committed:
Check non-existance after deleting meta but not saved yet

xristos3490 committed:
One more rename of note to payload

xristos3490 committed:
Introduce the email templates

xristos3490 committed:
Simplify and reorg templates and controllers

xristos3490 committed:
Lint fixes

xristos3490 committed:
Add plain templates

xristos3490 committed:
Handle plain mode in template functions

xristos3490 committed:
Lint fixes

xristos3490 committed:
Lint fixes

xristos3490 committed:
Reorg of files, renamed the email classes, and introduced a single entrypoint on the Controller class

xristos3490 committed:
Lint fixes

xristos3490 committed:
Lint fixes

xristos3490 committed:
revert unrelated changes

xristos3490 committed:
Rename email titles

xristos3490 committed:
Modify the database schema and fix the tests

xristos3490 committed:
Make linter happy

xristos3490 committed:
Merge branch 'mic/back-in-stock-notifications' into mic/bis-tables

xristos3490 committed:
Merge branch 'mic/bis-tables' into mic/bis-email-templates

xristos3490 committed:
Add the id at the batch processing index

xristos3490 committed:
Merge branch 'mic/bis-tables' into mic/bis-email-templates

xristos3490 committed:
Fix into->intro type

xristos3490 committed:
Rename custom tax location method

xristos3490 committed:
Fix verify template

xristos3490 committed:
Improve getter for formatted variations and add a unit test

xristos3490 committed:
Add unit tests for get_product_permalink/name

xristos3490 committed:
Rename emails id and classes for better semantics

xristos3490 committed:
Lint fixes on tests

xristos3490 committed:
Lint fixes on tests

xristos3490 committed:
Reorg namespaces into Emails, and move the templates into the email controller

xristos3490 committed:
Rename EmailManager, move two emails in transactional emails actions and modify trigger points

xristos3490 committed:
Lint

xristos3490 committed:
Fix customer data hook

xristos3490 committed:
Renamed confirm email to verified, and added better docblock for the customer data hook

xristos3490 committed:
Simplify and repurpose get_product_formatted_variation_list; handle emails rendering on the spot

xristos3490 committed:
Rename the template

xristos3490 committed:
Lint

xristos3490 committed:
Initial commit

xristos3490 committed:
Merge branch 'mic/back-in-stock-notifications' into wooprd-455

xristos3490 committed:
Refactor the notificationsProcessor and sync controller

xristos3490 committed:
Introduce the stock sync and async processor controllers and tests

xristos3490 committed:
Phpcs fixes

xristos3490 committed:
More lint fixes

xristos3490 committed:
More lint fixes

xristos3490 committed:
More lint fixes

xristos3490 committed:
More lint fixes

xristos3490 committed:
Repurpose StockManagementHelper and cleanups

xristos3490 committed:
Lint

xristos3490 committed:
Merge branch 'mic/back-in-stock-notifications' into wooprd-455

xristos3490 committed:
Prevent query for variations if no children

xristos3490 committed:
Rename the helper for better semantics

xristos3490 committed:
Fix AS args

xristos3490 committed:
Fix AS args in tests

xristos3490 committed:
Implement the attempt limit in the query

xristos3490 committed:
Add unit tests for the attempt limit on query

xristos3490 committed:
Cleanup debugs

xristos3490 committed:
Make the unit test for skipped a bit more specific

xristos3490 committed:
Fix parsing new AS args

xristos3490 committed:
Final changes for the new AS args structure

kalessil commented on plugins/woocommerce/src/Internal/StockNotifications/Utilities/StockManagementHelper.php:
```diff
@@ -0,0 +1,47 @@
+
+		$format           = array_fill( 0, count( $children ), '%d' );
+		$query_in         = '(' . implode( ',', $format ) . ')';
+		$managed_children = array_unique( $wpdb->get_col( $wpdb->prepare( "SELECT post_id FROM $wpdb->postmeta WHERE meta_key = '_manage_stock' AND meta_value != 'yes' AND post_id IN {$query_in}", $children ) ) ); // @codingStandardsIgnoreLine.
```
Using `DISTINCT` here will be, in most cases, better performing.

kalessil commented on plugins/woocommerce/src/Internal/StockNotifications/Utilities/StockManagementHelper.php:
```diff
@@ -0,0 +1,47 @@
+		$query_in         = '(' . implode( ',', $format ) . ')';
+		$managed_children = array_unique( $wpdb->get_col( $wpdb->prepare( "SELECT post_id FROM $wpdb->postmeta WHERE meta_key = '_manage_stock' AND meta_value != 'yes' AND post_id IN {$query_in}", $children ) ) ); // @codingStandardsIgnoreLine.
+
+		return array_map( 'intval', $managed_children );
```
I'm not 100% sure, but we might not need this casting - can we ensure the ids are not already ints?

kalessil commented on plugins/woocommerce/src/Internal/DataStores/StockNotifications/StockNotificationsDataStore.php:
```diff
@@ -492,4 +499,28 @@ function ( $result ) {
+		$query_in = '(' . implode( ',', $format ) . ')';
+		$sql      = $wpdb->prepare( // phpcs:ignore WordPress.DB.PreparedSQLPlaceholders.ReplacementsWrongNumber
+			"SELECT 1 FROM %i WHERE product_id IN $query_in AND status = %s LIMIT 1", // phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared
+			array_merge( array( $table ), $product_ids, array( NotificationStatus::ACTIVE ) )
```
There is a neat way to omit `array_merge` here: https://3v4l.org/O60BZ, in this context it would be:

```php
    array( $table, ...$product_ids, NotificationStatus::ACTIVE )
```

kalessil commented on plugins/woocommerce/src/Internal/StockNotifications/AsyncTasks/NotificationsProcessor.php:
```diff
@@ -0,0 +1,481 @@
+/**
+ * The async processor for sending stock notifications in bulk.
+ */
+class NotificationsProcessor {
```
Uh, this one is huge. Are there opportunities for decomposition here (dedicated services perhaps)?

xristos3490 replied:
Well, that's fair, @kalessil! I'll work on rearranging the responsibilities for this class and will re-request when I'm ready! Thanks!

kalessil commented on plugins/woocommerce/src/Internal/StockNotifications/Config.php:
```diff
@@ -0,0 +1,84 @@
+	/**
+	 * Get the supported product types.
+	 *
+	 * @return array
```
Looks like we can specify the array elements type here as well.

kalessil commented on plugins/woocommerce/src/Internal/StockNotifications/Config.php:
```diff
@@ -0,0 +1,84 @@
+		/**
+		 * Filter: woocommerce_stock_notifications_supported_product_types
+		 *
+		 * @since 0.0.0
```
Heh, we should probably update this line and the other two `@since` tags in this file.

kalessil commented on plugins/woocommerce/src/Internal/StockNotifications/Config.php:
```diff
@@ -0,0 +1,84 @@
+	/**
+	 * Get the supported product stock statuses.
+	 *
+	 * @return array
```
Looks like we can specify the array elements type here as well.

kalessil commented on plugins/woocommerce/src/Internal/StockNotifications/Config.php:
```diff
@@ -0,0 +1,84 @@
+	/**
+	 * Get the eligible stock statuses that trigger sending notifications.
+	 *
+	 * @return array
```
Looks like we can specify the array elements type here as well.

kalessil commented on plugins/woocommerce/src/Internal/StockNotifications/StockSyncController.php:
```diff
@@ -0,0 +1,146 @@
+			 *
+			 * @param int $product_id The product ID.
+			 */
+			do_action( 'woocommerce_stock_notifications_product_sync', $product_id );
```
Hint for consideration: this action could always accept an array of IDs for more convenience.

kalessil commented on plugins/woocommerce/src/Internal/StockNotifications/StockSyncController.php:
```diff
@@ -0,0 +1,146 @@
+	 * @param WC_Product|null $product      The product object (optional).
+	 * @return void
+	 */
+	public function handle_product_stock_status_change( $product_id, $stock_status, $product = null ) {
```
It looks like we can strictly type parameters here.

kalessil commented on plugins/woocommerce/src/Internal/StockNotifications/AsyncTasks/NotificationsProcessor.php:
```diff
@@ -0,0 +1,481 @@
+		$last_notified = $notification->get_date_notified();
+		$is_throttled  = $last_notified instanceof \WC_DateTime && $last_notified->getTimestamp() > ( time() - $threshold );
+
+		if ( $is_throttled ) {
```
Well, we can just `return $is_throttled` =)

kalessil commented on plugins/woocommerce/src/Internal/StockNotifications/AsyncTasks/NotificationsProcessor.php:
```diff
@@ -0,0 +1,481 @@
+				$is_sent = false;
+			}
+
+			if ( $is_sent ) {
```
Nit: optionally, this if-else could be part of the preceding try-catch.

kalessil commented on plugins/woocommerce/src/Internal/StockNotifications/StockSyncController.php:
```diff
@@ -0,0 +1,146 @@
+	public function handle_product_stock_status_change( $product_id, $stock_status, $product = null ) {
+		try {
+
+			if ( ! in_array( (string) $stock_status, Config::get_eligible_stock_statuses(), true ) ) {
```
Given it's a filter invocation under the hood, we might want to consider caching those statuses: e.g., the config methods themselves could support caching of values when they get called for the first time.

kalessil reviewed (commented):
Did the first review round: leaving some comments. Review was straightforward overall (thanks for good structure!) except the processor class, but I hope that on the next review round I'll do a deeper dive.

jorgeatorres commented on plugins/woocommerce/src/Internal/StockNotifications/AsyncTasks/NotificationsProcessor.php:
Just a general question: is there maybe an opportunity here to use [a `BatchProcessor`](https://github.com/woocommerce/woocommerce/blob/aeab61838c0be55f21d9aa7e08ef3c6fc4cad56f/plugins/woocommerce/src/Internal/BatchProcessing/BatchProcessorInterface.php#L13) here instead of adding yet another way to batch process things in the background. There might be some specific details that make this not viable for BIS, but I thought I'd ask.

It'd be good to consolidate and, if necessary, make improvements to our current library if that helps support other use cases. The current batch processor interface ([via `BatchProcessingController`](https://github.com/woocommerce/woocommerce/blob/a37132382f25f0d114157fc8b2ba187badf1e3a7/plugins/woocommerce/src/Internal/BatchProcessing/BatchProcessingController.php#L29)) is already used by the [HPOS sync process](https://github.com/woocommerce/woocommerce/blob/06d9edc4702446173a3d564c9eff0bf4e49b9410/plugins/woocommerce/src/Internal/DataStores/Orders/DataSynchronizer.php#L23), [HPOS cleanup](https://github.com/woocommerce/woocommerce/blob/bdf174af518b93ec913411b7e951077b42a38a41/plugins/woocommerce/src/Internal/DataStores/Orders/LegacyDataCleanup.php#L16), upgrade routines and also in Smooth Generator.

xristos3490 replied:
Hey @jorgeatorres!

The rationale for not selecting the BatchProcessor is that the goal here is to have a single AS task for each product that comes back in stock. The BatchProcessor is a more generic structure, and its inability to define AS arguments (managed via the interface) led us to create a new one heavily inspired by the current functionality of the extension.

That said, I don't mean we couldn't utilize this here. Here's a rough outline of how I'm thinking this:
- Every time a product comes in stock, add it to a FIFO queue of product IDs, and schedule the BatchProcessor (if it's not already scheduled).
- The BatchProcessor reads from the queue and handles notifications until the queue is empty.

This is a good alternative approach for the "NotificationsProcessor," but it does require robust handling of a FIFO queue, which likely means creating another database table for that. 

If I'm being honest, the BatchProcessor approach seems like a better plan to handle a large number of products in the same request. However, we didn't deviate too much from the extension because of:  
- The introduction of risk (due to being radically changed from the extension's workings), and  
- The fact that we can revisit and change how things work later.

Thoughts? Do you think it's worth it? I have mixed feelings, especially considering the feedback from @kalessil, who raised some questions about the size of that class.

jorgeatorres replied:
Hey @xristos3490! Thanks for getting back to me. A few thoughts:

> The BatchProcessor is a more generic structure, and its inability to define AS arguments (managed via the interface) [...]

This is correct, but also nothing prevents us from expanding on the batch processing infrastructure by adding support for, say, arguments. I might be missing something but if `BatchProcessor` supported even just one argument (a product ID), that would solve your use case.

I'm not saying these changes would've been super easy to implement (batch processors come usually from the container and there are more than a few other gotchas) but in my mind exploring that in the near future is definitely better than having two implementations of pretty much the same concept in core.

> [...] led us to create a new one heavily inspired by the current functionality of the extension.

I did see the resemblance which is in part why I asked. I think we could've prevented some duplication and also any efforts going into improving the current batch processing functionality would've benefited BIS too.

> This is a good alternative approach for the "NotificationsProcessor," but it does require robust handling of a FIFO queue, which likely means creating another database table for that.

Same as you, if we were to leave batch processing as is, I wouldn't suggest overcomplicating the implementation with a FIFO queue just to make use of it. I'm just (maybe naively) thinking that we can spend some time enhancing it to simplify things for users such as BIS.

> The fact that we can revisit and change how things work later.

This is fair, but I hope we don't forget about it (which I fear is likely the case). I think batch processors "with context" can be useful (not just for BIS) so having some common API and avoiding duplication would be nice.

> Thoughts? Do you think it's worth it? I have mixed feelings, especially considering the feedback from @kalessil, who raised some questions about the size of that class.

If making changes to batch processors to support your use case is too difficult, I don't think it's worth it and we can definitely revisit later. I understand the priority here is to land BIS in core, not to improve core's batch processing API, but it would've been nice if we had accomplished both.

A bit of a sidebar, but I think is important that we keep things like these in mind when merging any extension into core. IMHO it shouldn't just be about getting it to work but also see if there are any opportunities to re-use core APIs and/or improve core along the way.

As you said so yourself, the class is already a bit unwieldy, so I wouldn't suggest we add more complexity to it right now.

xristos3490 replied:
Thank you @jorgeatorres! Appreciate this feedback! 

I've made some changes to simplify the structure and separate concerns within the components, which should allow for more streamlined changes in the future.

As we discussed, for the BatchProcessor to work smoothly, we'll need support for context and a way to set intervals between batches to manage rate limits. Both are totally doable and deserve a dedicated workstream :)

xristos3490 committed:
Use spread op

xristos3490 committed:
Simplify throttle retun

xristos3490 committed:
Specify array elements

xristos3490 committed:
Rename and add DISTICT in query instead of array_unique

xristos3490 committed:
Merge branch 'mic/back-in-stock-notifications' into wooprd-455

xristos3490 committed:
Add runtime cache in config and rename hooks

xristos3490 committed:
Add the EligibilityService

xristos3490 committed:
Fix stocksynccontroller based on the new service

xristos3490 committed:
Fix tests for stocksync and minor fix in notificationtests

xristos3490 committed:
Use the aggregated hook in the processor

xristos3490 committed:
Refactor logic into smaller components

xristos3490 committed:
Lint fixes

xristos3490 committed:
More lint fixes

xristos3490 committed:
Merge branch 'mic/back-in-stock-notifications' into wooprd-455

xristos3490 committed:
minor rearrangement

xristos3490 committed:
Merge branch 'mic/back-in-stock-notifications' into wooprd-455

xristos3490 committed:
Fix unit tests of the async processor

xristos3490 committed:
Cleanup

xristos3490 committed:
Add new tables in WC_Install::get_tables()

xristos3490 committed:
Lint fixes

xristos3490 commented:
Hello folks! 👋 This is ready for review again :)

Sorry for the long delay. I experimented with reusing the BatchProcessor based on @jorgeatorres's feedback but ran into some challenges that prevented me from proceeding with this plan. If you're interested, I also tried a version of this using metadata on [this](https://github.com/woocommerce/woocommerce/pull/58728) PR.

@kalessil I've made some updates in this PR, mainly focusing on separating concerns between components and aiming for a simpler overall structure with smaller, self-contained services.

The refactor, in a nutshell, added:
- A new EligibilityService that will be used for various validation checks through the sync and background processes.
- A cycle state manager on the async task, acting like a data store (via wp options) for managing each product's metrics.
- A new job manager that handles AS tasks and reschedules.

@jimjasson Thanks for manually testing this, Jason! Given these changes, I recommend a retest to ensure no regressions. cc @PanosSynetos for capacity.

xristos3490 committed:
Remove deprecated service provider

kalessil commented on plugins/woocommerce/src/Internal/StockNotifications/Utilities/StockManagementHelper.php:
```diff
@@ -0,0 +1,55 @@
+			return array();
+		}
+
+		if ( isset( $this->managed_variations[ $product->get_id() ] ) ) {
```
Nit: `$product->get_id()` can be stored in a local variable to reduce the number of method invocations.

xristos3490 replied:
Thanks! Done in https://github.com/woocommerce/woocommerce/pull/58388/commits/47e217be60bd08142f730fedcf7f388f794f939e

kalessil commented on plugins/woocommerce/src/Internal/StockNotifications/AsyncTasks/JobManager.php:
```diff
@@ -0,0 +1,146 @@
+
+		try {
+
+			if ( WC()->queue()->get_next( self::AS_JOB_SEND_STOCK_NOTIFICATIONS, $args, self::AS_JOB_GROUP ) ) {
```
Nit: `WC()->queue()` is a good candidate for a private property initiated in the constructor.

xristos3490 replied:
Nice! 🙇 Done in https://github.com/woocommerce/woocommerce/pull/58388/commits/3f2b100ddd10b413f15e8d4687ba0f8da3fe30c4

kalessil commented on plugins/woocommerce/src/Internal/StockNotifications/AsyncTasks/NotificationsProcessor.php:
```diff
@@ -0,0 +1,258 @@
+		CycleStateService $cycle_state_service,
+		EmailManager $email_manager
+	): void {
+		$this->eligibility_service = $eligibility_service;
```
The populated properties here can be potentially declared with types (if our min PHP level supports this ofcourse - need verification).

xristos3490 replied:
Nice catch! I have added them in other interfaces already 😅 

[PHP 7.4 added typed properties](https://www.php.net/manual/en/language.oop5.properties.php), including custom class types, which works fine for us since we require 7.4+. The EmailEditor already uses this feature.

Done in https://github.com/woocommerce/woocommerce/pull/58388/commits/be51cbdca38fb7ebb6f9fc1d601e15c93daeb38d

kalessil reviewed (approved):
LGTM (pre-approving, there are some unresolved nits but it's up to ya folks to followup/ignore those)! I completed the final review round:
- the latest version of the class structure looks well broken down and refactoring-friendly
- no performance concerns related to DB queries

As a side note, when we have an RC state for this project, it might be helpful to stress-test it to ensure we have no implementation-level bottlenecks.

xristos3490 committed:
Add typed class props on processor

xristos3490 committed:
Polish sync hook

xristos3490 committed:
Cache queue instance into prop

xristos3490 committed:
Use local product_id ref

xristos3490 committed:
harmonize logs group

xristos3490 committed:
Oops

jimjasson reviewed (commented):
Hey @xristos3490, 

I did another manual test after the recent changes!

<details><summary>Details</summary>
<p>
### WordPress Environment ###

WordPress address (URL): http://wp.local
Site address (URL): http://wp.local
WC Version: 10.1.0-dev
Legacy REST API Package Version: The Legacy REST API plugin is not installed on this site.
Action Scheduler Version: ✔ 3.9.2
Log Directory Writable: ✔
WP Version: 6.8.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: nginx/1.26.1
Server Architecture: Darwin 23.3.0 arm64
PHP Version: 8.2.27
PHP Post Max Size: 1,000 MB
PHP Time Limit: 1200
PHP Max Input Vars: 4000
cURL Version: 8.4.0
(SecureTransport) LibreSSL/3.3.6

SUHOSIN Installed: –
MySQL Version: 8.0.35
Max Upload Size: 300 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 10.1.0
WC Database Prefix: wp_
Total Database Size: 43.12MB
Database Data Size: 25.82MB
Database Index Size: 17.30MB
wp_woocommerce_sessions: Data: 1.44MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.14MB + Index: 0.11MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 1.50MB + Index: 0.84MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 1.02MB + Index: 0.64MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 1.03MB + Index: 0.33MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.19MB + Index: 0.25MB + Engine InnoDB
wp_gf_draft_submissions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_gf_entry: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_gf_entry_meta: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_gf_entry_notes: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_gf_form: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_gf_form_meta: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_gf_form_revisions: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_gf_form_view: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 7.02MB + Index: 0.16MB + Engine InnoDB
wp_postmeta: Data: 6.48MB + Index: 8.63MB + Engine InnoDB
wp_posts: Data: 2.39MB + Index: 0.66MB + Engine InnoDB
wp_snippets: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.11MB + Index: 0.06MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.31MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.06MB + Index: 0.02MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_wc_booking_relationships: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_bookings_availability: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_bookings_availabilitymeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_mnm_child_items: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_addresses: Data: 0.14MB + Index: 0.25MB + Engine InnoDB
wp_wc_order_bundle_lookup: Data: 0.05MB + Index: 0.09MB + Engine InnoDB
wp_wc_order_composite_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_operational_data: Data: 0.09MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.11MB + Index: 0.20MB + Engine InnoDB
wp_wc_order_stats: Data: 0.08MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_orders: Data: 0.13MB + Index: 0.17MB + Engine InnoDB
wp_wc_orders_meta: Data: 1.47MB + Index: 1.94MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.09MB + Index: 0.06MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.20MB + Index: 0.52MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_stock_notificationmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_stock_notifications: Data: 0.41MB + Index: 0.41MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wcs_payment_retries: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_bis_activity: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_woocommerce_bis_notifications: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_woocommerce_bis_notificationsmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_bundled_itemmeta: Data: 0.06MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_bundled_items: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_gc_activity: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_gc_cards: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_gc_cardsmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_prl_amplifier_cache: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_woocommerce_prl_deploymentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_prl_deployments: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_prl_frequencies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_prl_generator_queue: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_prl_tracking_conversions: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wt_iew_action_history: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wt_iew_cron: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wt_iew_ftp: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wt_iew_mapping_template: Data: 0.02MB + Index: 0.00MB + Engine InnoDB

### Post Type Counts ###

attachment: 52
bookable_resource: 4
csp_rules: 1
global_product_addon: 1
page: 19
post: 3
prl_engine: 1
product: 199
product_variation: 76
revision: 101
shop_coupon: 7
shop_order: 375
shop_order_refund: 1
shop_subscription: 57
wc_booking: 2392
wc_membership_plan: 1
wc_user_membership: 1
wishlist: 1
wp_font_face: 33
wp_font_family: 12
wp_global_styles: 4
wp_navigation: 1
wp_template: 5
wp_template_part: 3
yay-currency-manage: 2

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ✔

### Active Plugins (11) ###

PU2 DevTools: by  –
Query Monitor: by John Blackbourn – 3.16.4
WooCommerce Back In Stock Notifications: by Woo – 3.1.0
WooCommerce Bookings: by WooCommerce – 2.2.7
WooCommerce Composite Products: by Woo – 10.6.0
WooCommerce Gift Cards: by Woo – 2.6.0
WooCommerce Product Add-Ons: by Woo – 7.9.0
WooCommerce Product Bundles: by Woo – 8.4.1
WooCommerce Product Recommendations: by Woo – 4.1.3
WooCommerce Subscriptions: by WooCommerce – 7.3.1
WooCommerce: by Automattic – 10.1.0-dev

### Inactive Plugins (45) ###

Advanced Cron Manager: by BracketSpace – 2.5.3
Akismet Anti-spam: Spam Protection: by Automattic - Anti-spam Team – 5.3.1
Classic Widgets: by WordPress Contributors – 0.3
Code Snippets: by Code Snippets Pro – 3.6.5.1
Composite Products - Conditional Images: by WooCommerce – 2.0.1
Core Rollback: by Andy Fragen – 1.3.5
Facebook for WooCommerce: by Facebook – 3.4.8
Gift Cards - Coupon Input: by WooCommerce – 2.0.1
Gravity Forms: by Gravity Forms – 2.8.7
Import Export Suite for WooCommerce: by WebToffee – 1.3.0
Loco Translate: by Tim Whitlock – 2.6.14
Perfect Brands WooCommerce: by QuadLayers – 3.4.8
Product Bundles - Bulk Discounts: by WooCommerce – 2.0.1
Product Bundles - Variation Bundles: by WooCommerce – 2.0.2
Product Recommendations - Custom Locations: by WooCommerce – 2.0.4
Redis Object Cache: by Till Krüss – 2.5.4
Role Based Pricing for WooCommerce: by Addify – 2.2.0
Taxonomy Switcher: by WebDevStudios – 1.0.7
Wholesale For WooCommerce: by wpexpertsio – 2.3.1
WooCommerce All Products for Subscriptions: by Woo – 6.0.6
WooCommerce Analytics: by WooCommerce – 0.0.1
WooCommerce Bookings: by WooCommerce – 2.0.0
WooCommerce Bookings: by WooCommerce – 1.15.72
WooCommerce Bookings: by WooCommerce – 1.16.0
WooCommerce Brands: by WooCommerce – 1.7.4
WooCommerce Conditional Shipping and Payments: by Woo – 2.2.0
WooCommerce CyberSource Gateway: by SkyVerge – 2.9.0
WooCommerce Dummy Payments Gateway: by Woo – 1.10.0
WooCommerce FedEx Shipping: by WooCommerce – 4.3.1
WooCommerce Memberships: by SkyVerge – 1.26.11
WooCommerce Min/Max Quantities: by Woo – 5.2.4
WooCommerce Mix and Match Products: by Kathy Darling – 2.7.8
WooCommerce Name Your Price: by Kathy Darling – 3.6.0
WooCommerce PayPal Payments: by PayPal – 3.0.5 (update to version 3.0.7 is available)
WooCommerce Pre-Orders: by WooCommerce – 2.1.4
WooCommerce Print Invoices/Packing Lists: by SkyVerge – 3.13.8
WooCommerce Product Bundles - Top Add to Cart Button: by SomewhereWarm – 1.0.3
WooCommerce Product Dependencies: by WooCommerce – 2.0.1
WooCommerce Smooth Generator: by Automattic – 1.2.0
WooCommerce Subscriptions Core: by Automattic – 7.6.0
WooCommerce WishLists: by Element Stark – 2.3.0
WooPayments: by WooCommerce – 7.8.1 (update to version 9.5.0 is available)
WordPress Beta Tester: by WordPress Upgrade/Install Team – 3.5.5
WordPress Importer: by wordpressdotorg – 0.8.4
YayCurrency Pro: by YayCommerce – 2.4.4.2

### Dropin Plugins () ###

db.php: Query Monitor Database Class (Drop-in)

### Settings ###

Legacy API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: right_space
Thousand Separator: .
Decimal Separator: ,
Number of Decimals: 2
Taxonomies: Product Types: booking (booking)
bundle (bundle)
composite (composite)
external (external)
grouped (grouped)
mix and match (mix-and-match)
simple (simple)
subscription (subscription)
variable (variable)
variable subscription (variable-subscription)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: ✔
Enforce Approved Product Download Directories: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: ✔
Enabled Features: analytics
marketplace
order_attribution
hpos_fts_indexes
remote_logging
email_improvements
blueprint
point_of_sale
custom_order_tables


### Logging ###

Enabled: ✔
Handler: Automattic\WooCommerce\Internal\Admin\Logging\LogHandlerFileV2
Retention period: 30 days
Level threshold: –
Log directory size: 7 MB

### WC Pages ###

Shop base: #7 - /shop/
Cart: #8 - /cart/ -  Contains the woocommerce/cart block
Checkout: #9 - /checkout/ -  Contains the woocommerce/checkout block
My account: #10 - /my-account/ -  Contains the [woocommerce_my_account] shortcode
Terms and conditions: ❌ Page not set

### Theme ###

Name: Storefront
Version: 4.6.1
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
Theme type: Classic theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Subscriptions ###

WCS_DEBUG: ✔ No
Subscriptions Mode: ✔ Live
Subscriptions Live URL: http://wp.local
Subscriptions-core Library Version: 8.1.1
Subscription Statuses: wc-active: 36
wc-on-hold: 15
wc-pending: 6

WooCommerce Account Connected: ✔ Yes
Active Product Key: ❌ No
Custom Retry Rules: ✔ No
Custom Retry Rule Class: ✔ No
Custom Raw Retry Rule: ✔ No
Custom Retry Rule: ✔ No
Retries Migration Status: ✔ Completed
Report Cache Enabled: ✔ Yes
Cache Update Failures: ✔ 0 failure

### Store Setup ###

Country / State: United States (US) — California

### Subscriptions by Payment Gateway ###

other: wc-active: 1
wc-on-hold: 3

Check payments: wc-on-hold: 1

### Payment Gateway Support ###

Direct bank transfer: products
Check payments: products
Cash on delivery: products

### Back In Stock ###

Database Version: 3.1.0
Loopback Test: ✔

### Composite Products ###

Database Version: 10.6.0
Loopback Test: ✔
Template Overrides: –

### Gift Cards ###

Database Version: 2.6.0
Loopback Test: ✔
Task Queueing Test: ✔

### Product Bundles ###

Database Version: 8.4.1
Loopback Test: ✔
Template Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customize-store
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
onboarding
onboarding-tasks
pattern-toolkit-full-composability
product-custom-fields
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
printful
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page
launch-your-store
add-to-cart-with-options-stepper-layout

Disabled Features: product-data-views
experimental-blocks
experimental-iapi-mini-cart
experimental-iapi-runtime
coming-soon-newsletter-template
minified-js
product-pre-publish-modal
settings
async-product-editor-category-field
product-editor-template-system
use-wp-horizon

Daily Cron: ✔ Next scheduled: 2025-07-02 11:18:44 +03:00
Options: ✔
Notes: 128
Onboarding: completed

### Action Scheduler ###

Canceled: 3
Oldest: -0001-11-30 00:00:00 +0000
Newest: -0001-11-30 00:00:00 +0000

Complete: 1,269
Oldest: 2025-06-04 10:07:28 +0000
Newest: 2025-07-02 06:18:14 +0000

Failed: 411
Oldest: 2024-02-26 11:21:02 +0000
Newest: 2025-06-24 07:25:38 +0000

Pending: 8
Oldest: 2025-07-02 06:18:23 +0000
Newest: 2025-07-19 07:46:31 +0000


### Product Recommendations ###

Database Version: 4.1.3
Loopback Test: ✔
Task Queueing Test: ✔
Page Cache Test: No cache detected

### Status report information ###

Generated at: 2025-07-02 09:18:21 +03:00

</p>
</details> 


(1) **Out of Stock to In Stock Simple Product**

I created 10 notifications for an Out of Stock Simple Product and then changed its status to In Stock. 

An AS action with the product ID as an argument was created:

![Markup on 2025-07-02 at 09:14:40](https://github.com/user-attachments/assets/f3c69195-5afc-44b6-ac4b-765af5632360)

A log entry was added:

![Markup on 2025-07-02 at 09:16:44](https://github.com/user-attachments/assets/abd3ee51-738d-4198-a715-0f98ed0a5a45)

After the AS action was completed, 10 emails were sent. 

⚠️ However, the log wasn't updated as you mention in the testing instructions:

> INFO: Completed cycle for product 123. Sent: 15, Skipped: 2, Failed: 0, Duration: 8 seconds. Total notifications processed: 17

Can you confirm that this is expected?

❌ When making the product in stock, then an admin notice shows up, but it points to a 404 page.

Maybe the redirect URL should be `admin.php?page=wc-customer-stock-notifications` instead of `wp-admin/admin.php?page=customer-stock-notifications`?

![Markup on 2025-07-02 at 09:25:00](https://github.com/user-attachments/assets/853f1cf9-4523-4a45-b710-2bacbaada217)

(2) **Out of Stock to On Backorder Simple Product**

I created 1 notification for an Out of Stock Simple Product and then changed its status to On Backorder. 

Same as in (1), notification e-mail sent correctly. 

(3) **0 stock to In Stock Simple Product**

I created 1 notification for a Simple Product with 0 stock and then changed its stock to 10. 

Same as in (1), notification e-mail sent correctly. 

(4) **Error handling: Product goes back to out of stock before sending notification**

I created 1 notification for an Out of Stock Simple Product and then changed its status to In Stock.

Before the notification got sent, I changed its status back to Out of Stock. 

Notifications are not sent and error was logged:

![Markup on 2025-07-02 at 09:34:22](https://github.com/user-attachments/assets/a56b4b6e-a0f6-4f5a-b777-f92fdc0bf987)

(5) **Error handling: Product becomes Private before sending notification**

I created 1 notification for an Out of Stock Simple Product and then changed its status to In Stock.

Before the notification got sent, I changed it to Private.

⚠️ Notifications were sent. Can you confirm that this is expected?

(6) **Error handling: Product becomes Draft before sending notification**

I created 1 notification for an Out of Stock Simple Product and then changed its status to In Stock.

Before the notification got sent, I changed it to Draft.

⚠️ Notifications were sent. Can you confirm that this is expected?

(7) **Error handling: Product is deleted before sending notification**

I created 1 notification for an Out of Stock Simple Product and then changed its status to In Stock.

Before the notification got sent, I deleted the product.

⚠️ Notifications were sent. Can you confirm that this is expected?

(8) **Out of stock variation to In Stock**

I created a Variable Product with a variation that is Out of stock. I created a notification using the variation ID. I changed the variation's stock status to In Stock. 

The notification e-mail was sent correctly. 

(9) **Variation with 0 stock to In Stock**

I created a Variable Product with a variation that is managing stock and has 0 remaining stock. I created a notification using the variation ID. I increased the variation's stock to 10.

The notification e-mail was sent correctly. 

(10) **Variable Product with Stock Management and notification with the parent ID**

I created a Variable Product with stock management enabled and set its stock to 0. I created a notification using the Variable Product ID. I increased its stock to 10. 

I received a single e-mail for the parent product, but AS tasks were processed both for the parent product and every variation that isn't managing stock. 

(11) **Variable Product with Stock Management and notification with the variation ID**

I created a Variable Product with stock management enabled and set its stock to 0. I created a variation without stock management and a notification using the variation ID. I increased the parent stock to 10. 

I received a single e-mail for the variation product, but AS tasks were processed both for the parent product and every variation that isn't managing stock.

xristos3490 committed:
Harmonize the logs group

xristos3490 committed:
Properly handle unanavailable products

xristos3490 committed:
Add another test for trashed products

xristos3490 committed:
Fix the notice URL

xristos3490 commented:
Thanks for the review, Jason! 🙇 

> ⚠️ However, the log wasn't updated as you mention in the testing instructions:

Thanks! I accidentally and last-minute changed the logs group and left the JobManager's source untouched 😅 Everything should be now available under the `wc-customer-stock-notifications` log. Fixed in https://github.com/woocommerce/woocommerce/pull/58388/commits/788033342b5ff164de509d2cc619d3488f3874af

> ❌ When making the product in stock, then an admin notice shows up, but it points to a 404 page.

Thanks, fixed in https://github.com/woocommerce/woocommerce/pull/58388/commits/e0ffdde6a2356c7b89725ca376239cb4b1548839

> (5) Error handling: Product becomes Private before sending notification
> (6) Error handling: Product becomes Draft before sending notification

Notifications sent by store managers and admins should be received, so if you used test notifications with your active Woo user, this is expected. Could you double-check on this?

> (7) Error handling: Product is deleted before sending notification

Nice catch! Thanks! Apparently, this falls under the same criteria for admin or store managers to receive notifications even when the product isn't "active." Fixed in https://github.com/woocommerce/woocommerce/pull/58388/commits/729a8833c433a3d5e60b45f66af184a8dcefe9d9 and added another unit test for this in https://github.com/woocommerce/woocommerce/pull/58388/commits/d6330ce32050ffc9fad2abba25a3358d1169564b

> I received a single e-mail for the parent product, but AS tasks were processed both for the parent product and every variation that isn't managing stock.

This isn't expected, and I cannot replicate it either. Can you share a few more details on how you set up your managed variations?

jimjasson commented:
> Everything should be now available under the wc-customer-stock-notifications log.

Confirmed!

> Thanks, fixed in https://github.com/woocommerce/woocommerce/commit/e0ffdde6a2356c7b89725ca376239cb4b1548839

Confirmed, that it has been resolved. 

> Notifications sent by store managers and admins should be received, so if you used test notifications with your active Woo user, this is expected. Could you double-check on this?

Yes, looks like some of my tests were with an admin users. Confirmed that notifications are not sent in these scenarios for non admin users. All good!

> Nice catch! Thanks! Apparently, this falls under the same criteria for admin or store managers to receive notifications even when the product isn't "active."

I confirmed that the notification is not sent for the deleted product, but the log doesn't correctly reflect the failure:

![Markup on 2025-07-02 at 15:40:20](https://github.com/user-attachments/assets/e0cf0b54-e514-4397-9559-13aa640e92d5)

⚠️ Shouldn't Skipped or Failed be 1 in this case?

> This isn't expected, and I cannot replicate it either. Can you share a few more details on how you set up your managed variations?

Discussed via call, but this is expected -- it was the wording that made it confusing. I was referring to variations with Manage Stock? enabled, not disabled.

xristos3490 commented:
Thanks for re-testing, @jimjasson  :) 

> ⚠️ Shouldn't Skipped or Failed be 1 in this case?

Yup, this is expected. It's not a failure, but a skipped one. Failures during this process refer to technical issues that prevented the email from sending - at least, as much as we can control. In my opinion, ideally when a product gets deleted, all notifications related to this should cascade. However, that can be a future improvement if necessary.

> Discussed via call, but this is expected -- it was the wording that made it confusing. I was referring to variations with Manage Stock? enabled, not disabled.

Thanks! Yes, I understood you were talking about managed variations _by the parent_. What you outlined in your initial review is accurate and as expected!

xristos3490 commented:
@jimjasson Correction on the "skipped/failed" part - What is really happening here is that the product is not eligible for notifications when trashed, so the process ends without processing any notifications. This is better in case this happened by mistake and you won't need to go through a large list of notifications to re-activate them just to resend.

Hope this makes sense :)

jimjasson reviewed (approved):
Thank you for addressing all of my review comments and explaining some grey areas! I fully tested this feature and everything works as expected. 

As discussed over our call, the product permalinks for the Notifications list table are not correct, but this is not related to this PR and should be resolved separately.
